### DISCUSSION TITLE
Is it abuse to use the restart-* configs for cron-like functions?

### DISCUSSION THREAD

**Discussion by ghost** [davmac314/dinit#421] [Q&A] [answered]
See this service-file:

    # This is not long-running and will quickly exit
    command = echo hello world

    # Restart on exit
    restart = true
    # ... after 5 minutes
    restart-delay = 300
    # ... until machine shutdown
    restart-limit-count = 0
    # ... and have no collateral damage
    smooth-recovery = true

    # This is really "scripted" pretending to be a process
    type = process

Is this considered an abuse of dinit service-management?
Are there any mechanisms that make this very costly?

**crond(1) is not an option**, and I am also curious.

**Comment by davmac314** [comment, marked as answer]
I mean it's definitely a hack and not "sanctioned" as such but it will certainly work. There's nothing costly about it.

**Reply by ghost** [reply]
If possible, I'd like to have a discussion about this. I believe that it is a valid and under-served use-case. This is discussion about theory so I may move it to another thread but I am curious on your perspective.

---

**The middle ground between a "oneshot" and a process, should be the job of the service-manager**.
As you have stated, this is already supported by `dinit` - but I think it should also be sanctioned. 

After sitting on this for a while, I think that `crond` is redundant for most users. It is just not a good solution:

- *Service* managers should be responsible for *service* restarts, and should ***not*** relegate this to cron !!!
- Majority of machines (embedded, laptops, and phones) do not have continuous uptime, and cron cannot be relied upon
- Extremely wide problem-scope, and no two cron implementations agree on what it should actually do

This is all on top of that, in modern day, cron just does not make sense. Can only be run by root? User-services exist. Need a SUID-bit program for configuration? User-services exist. Need complex workarounds to get environment variables? User-services exist. Need complex workarounds to handle multiple timezones? User-services (and TZ=) exist.

I initially thought `anacron` was redundant software, but I now think they had it mostly correct. The only job of cron should be to ***audit*** that something ran, and if it did not, then respond (flag, script, trigger, etc). **Everything else is the service-manager's job**.

Dinit should have `restart = on-success`, which restarts a `type = scripted` on success.
Or have `type = loop`, which is a oneshot that will run again on success.

**Reply by ghost** [reply]
To clarify, the following service file is what (I think) cron should be:

    # anacron-like program that can handle seconds, minutes, and hours.
    command = ocron

    # Run every minute
    restart-delay = 60

    # Oneshot that will always be restarted. Do not restart dependents unless it fails.
    type = loop

**OR**


    # anacron-like program that can handle seconds, minutes, and hours.
    command = ocron

    # Restart even on success (in practice, only affects type=scripted)
    restart = on-success
    # ... after 1 minute
    restart-delay = 60
    # ... until machine shutdown
    restart-limit-count = 0
    # ... and have no collateral damage
    smooth-recovery = true

    type = scripted

This takes control of this service type (middle ground between oneshot and process) away from `crond` and gives it back to the service-manager, is resilient to inconsistent uptime, and clearly defines the role of cron.

This has overall been a rant, but as I said prior I am curious on your thoughts of this.

**Reply by davmac314** [reply]
I haven't put much thought into it yet (and probably won't at least until after 1.0 is released).

There is an [open issue](https://github.com/davmac314/dinit/issues/115).

My only thoughts are: yes, it's probably a good idea to have some kind of explicit support for periodic one-shot services, but waking up a process every minute just in case there is something to do is antithetical to Dinit's philosophy (and is probably bad for laptop battery life for example).
